samza-core/src/main/scala/org/apache/samza/system/SSPMetadataCache.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.system;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.samza.util.Clock;
+
+
+/**
+ * Fetches and caches metadata about a set of SSPs. This can prefetch the metadata for a defined set of SSPs when it is
+ * fetching metadata for some arbitrary SSP.
+ */
+public class SSPMetadataCache {
```

Comment:
Although there are some similarities in functionality, it will be difficult to have a shared class at this point:
1. StreamMetadataCache has a "partitionsMetadataOnly" parameter which is specific to stream metadata.
2. SSPMetadataCache uses prefetching, but StreamMetadataCache does not, so we still need some extra logic.
3. The code structure of SSPMetadataCache is a bit different from StreamMetadataCache, due to the prefetching. It might be possible to do some refactoring to get them more similar, but given the two other reasons, it might not be worth it.